File: tests/test_types.cpp
/**
 * Test: test_types
 * Tests LLVM Type creation and inspection
 *
 * LLVM-C APIs covered:
 * - LLVMInt1TypeInContext(), LLVMInt8TypeInContext(), LLVMInt16TypeInContext()
 * - LLVMInt32TypeInContext(), LLVMInt64TypeInContext(),
 * LLVMInt128TypeInContext()
 * - LLVMIntTypeInContext()
 * - LLVMHalfTypeInContext(), LLVMFloatTypeInContext(),
 * LLVMDoubleTypeInContext()
 * - LLVMVoidTypeInContext()
 * - LLVMPointerTypeInContext()
 * - LLVMArrayType2()
 * - LLVMVectorType()
 * - LLVMFunctionType()
 * - LLVMStructTypeInContext(), LLVMStructCreateNamed(), LLVMStructSetBody()
 * - LLVMPrintTypeToString(), LLVMGetTypeKind()
 * - LLVMGetIntTypeWidth()
 * - LLVMTypeIsSized()
 */

#include <cstdio>
#include <llvm-c/Analysis.h>
#include <llvm-c/Core.h>

const char *type_kind_name(LLVMTypeKind kind) {
  switch (kind) {
  case LLVMVoidTypeKind:
    return "void";
  case LLVMHalfTypeKind:
    return "half";
  case LLVMFloatTypeKind:
    return "float";
  case LLVMDoubleTypeKind:
    return "double";
  case LLVMX86_FP80TypeKind:
    return "x86_fp80";
  case LLVMFP128TypeKind:
    return "fp128";
  case LLVMPPC_FP128TypeKind:
    return "ppc_fp128";
  case LLVMLabelTypeKind:
    return "label";
  case LLVMIntegerTypeKind:
    return "integer";
  case LLVMFunctionTypeKind:
    return "function";
  case LLVMStructTypeKind:
    return "struct";
  case LLVMArrayTypeKind:
    return "array";
  case LLVMPointerTypeKind:
    return "pointer";
  case LLVMVectorTypeKind:
    return "vector";
  case LLVMMetadataTypeKind:
    return "metadata";
  case LLVMTokenTypeKind:
    return "token";
  case LLVMScalableVectorTypeKind:
    return "scalable_vector";
  case LLVMBFloatTypeKind:
    return "bfloat";
  case LLVMX86_AMXTypeKind:
    return "x86_amx";
  case LLVMTargetExtTypeKind:
    return "target_ext";
  default:
    return "unknown";
  }
}

int main() {
  LLVMContextRef ctx = LLVMContextCreate();
  LLVMModuleRef mod = LLVMModuleCreateWithNameInContext("test_types", ctx);

  // Integer types
  LLVMTypeRef i1 = LLVMInt1TypeInContext(ctx);
  LLVMTypeRef i8 = LLVMInt8TypeInContext(ctx);
  LLVMTypeRef i16 = LLVMInt16TypeInContext(ctx);
  LLVMTypeRef i32 = LLVMInt32TypeInContext(ctx);
  LLVMTypeRef i64 = LLVMInt64TypeInContext(ctx);
  LLVMTypeRef i128 = LLVMInt128TypeInContext(ctx);
  LLVMTypeRef i256 = LLVMIntTypeInContext(ctx, 256);

  // Floating point types
  LLVMTypeRef f16 = LLVMHalfTypeInContext(ctx);
  LLVMTypeRef bf16 = LLVMBFloatTypeInContext(ctx);
  LLVMTypeRef f32 = LLVMFloatTypeInContext(ctx);
  LLVMTypeRef f64 = LLVMDoubleTypeInContext(ctx);

  // Void type
  LLVMTypeRef void_ty = LLVMVoidTypeInContext(ctx);

  // Pointer type (opaque pointer)
  LLVMTypeRef ptr = LLVMPointerTypeInContext(ctx, 0);

  // Array type
  LLVMTypeRef arr_i32_10 = LLVMArrayType2(i32, 10);

  // Vector type
  LLVMTypeRef vec_i32_4 = LLVMVectorType(i32, 4);

  // Function type: i32(i32, i32)
  LLVMTypeRef func_params[] = {i32, i32};
  LLVMTypeRef func_ty = LLVMFunctionType(i32, func_params, 2, 0);

  // Function type with varargs: i32(i32, ...)
  LLVMTypeRef vararg_params[] = {i32};
  LLVMTypeRef vararg_func_ty = LLVMFunctionType(i32, vararg_params, 1, 1);

  // Anonymous struct type: {i32, f64}
  LLVMTypeRef struct_elems[] = {i32, f64};
  LLVMTypeRef anon_struct = LLVMStructTypeInContext(ctx, struct_elems, 2, 0);

  // Packed anonymous struct type: <{i8, i32}>
  LLVMTypeRef packed_elems[] = {i8, i32};
  LLVMTypeRef packed_struct = LLVMStructTypeInContext(ctx, packed_elems, 2, 1);

  // Named struct type
  LLVMTypeRef named_struct = LLVMStructCreateNamed(ctx, "MyStruct");
  LLVMTypeRef named_elems[] = {i32, ptr, f64};
  LLVMStructSetBody(named_struct, named_elems, 3, 0);

  // Opaque struct (no body set)
  LLVMTypeRef opaque_struct = LLVMStructCreateNamed(ctx, "OpaqueStruct");

  // Add global variables to make types visible in output
  LLVMAddGlobal(mod, i32, "global_i32");
  LLVMAddGlobal(mod, arr_i32_10, "global_arr");
  LLVMAddGlobal(mod, vec_i32_4, "global_vec");
  LLVMAddGlobal(mod, anon_struct, "global_anon_struct");
  LLVMAddGlobal(mod, packed_struct, "global_packed_struct");
  LLVMAddGlobal(mod, named_struct, "global_named_struct");

  // Add a function declaration to show function type
  LLVMAddFunction(mod, "example_func", func_ty);
  LLVMAddFunction(mod, "example_vararg_func", vararg_func_ty);

  // Verify module
  char *error = nullptr;
  if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
    fprintf(stderr, "; Verification failed: %s\n", error);
    LLVMDisposeMessage(error);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);
    return 1;
  }
  LLVMDisposeMessage(error);

  // Print diagnostic comments
  printf("; Test: test_types\n");
  printf(";\n");
  printf("; Integer types:\n");
  printf(";   i1 width: %u, kind: %s\n", LLVMGetIntTypeWidth(i1),
         type_kind_name(LLVMGetTypeKind(i1)));
  printf(";   i8 width: %u, kind: %s\n", LLVMGetIntTypeWidth(i8),
         type_kind_name(LLVMGetTypeKind(i8)));
  printf(";   i16 width: %u, kind: %s\n", LLVMGetIntTypeWidth(i16),
         type_kind_name(LLVMGetTypeKind(i16)));
  printf(";   i32 width: %u, kind: %s\n", LLVMGetIntTypeWidth(i32),
         type_kind_name(LLVMGetTypeKind(i32)));
  printf(";   i64 width: %u, kind: %s\n", LLVMGetIntTypeWidth(i64),
         type_kind_name(LLVMGetTypeKind(i64)));
  printf(";   i128 width: %u, kind: %s\n", LLVMGetIntTypeWidth(i128),
         type_kind_name(LLVMGetTypeKind(i128)));
  printf(";   i256 width: %u, kind: %s\n", LLVMGetIntTypeWidth(i256),
         type_kind_name(LLVMGetTypeKind(i256)));
  printf(";\n");
  printf("; Floating point types:\n");
  printf(";   half kind: %s\n", type_kind_name(LLVMGetTypeKind(f16)));
  printf(";   bfloat kind: %s\n", type_kind_name(LLVMGetTypeKind(bf16)));
  printf(";   float kind: %s\n", type_kind_name(LLVMGetTypeKind(f32)));
  printf(";   double kind: %s\n", type_kind_name(LLVMGetTypeKind(f64)));
  printf(";\n");
  printf("; Other types:\n");
  printf(";   void kind: %s, sized: %s\n",
         type_kind_name(LLVMGetTypeKind(void_ty)),
         LLVMTypeIsSized(void_ty) ? "yes" : "no");
  printf(";   pointer kind: %s, sized: %s\n",
         type_kind_name(LLVMGetTypeKind(ptr)),
         LLVMTypeIsSized(ptr) ? "yes" : "no");
  printf(";   array kind: %s, sized: %s\n",
         type_kind_name(LLVMGetTypeKind(arr_i32_10)),
         LLVMTypeIsSized(arr_i32_10) ? "yes" : "no");
  printf(";   vector kind: %s, sized: %s\n",
         type_kind_name(LLVMGetTypeKind(vec_i32_4)),
         LLVMTypeIsSized(vec_i32_4) ? "yes" : "no");
  printf(";   function kind: %s, sized: %s\n",
         type_kind_name(LLVMGetTypeKind(func_ty)),
         LLVMTypeIsSized(func_ty) ? "yes" : "no");
  printf(";\n");
  printf("; Struct types:\n");
  printf(";   anon_struct kind: %s, packed: %s\n",
         type_kind_name(LLVMGetTypeKind(anon_struct)),
         LLVMIsPackedStruct(anon_struct) ? "yes" : "no");
  printf(";   packed_struct kind: %s, packed: %s\n",
         type_kind_name(LLVMGetTypeKind(packed_struct)),
         LLVMIsPackedStruct(packed_struct) ? "yes" : "no");

  const char *named_struct_name = LLVMGetStructName(named_struct);
  printf(";   named_struct name: %s, opaque: %s\n", named_struct_name,
         LLVMIsOpaqueStruct(named_struct) ? "yes" : "no");

  const char *opaque_struct_name = LLVMGetStructName(opaque_struct);
  printf(";   opaque_struct name: %s, opaque: %s\n", opaque_struct_name,
         LLVMIsOpaqueStruct(opaque_struct) ? "yes" : "no");

  // Print type strings
  printf(";\n");
  printf("; Type strings:\n");
  char *i32_str = LLVMPrintTypeToString(i32);
  printf(";   i32: %s\n", i32_str);
  LLVMDisposeMessage(i32_str);

  char *arr_str = LLVMPrintTypeToString(arr_i32_10);
  printf(";   [10 x i32]: %s\n", arr_str);
  LLVMDisposeMessage(arr_str);

  char *func_str = LLVMPrintTypeToString(func_ty);
  printf(";   func type: %s\n", func_str);
  LLVMDisposeMessage(func_str);

  printf("\n");

  // Print module IR
  char *ir = LLVMPrintModuleToString(mod);
  printf("%s", ir);
  LLVMDisposeMessage(ir);

  // Cleanup
  LLVMDisposeModule(mod);
  LLVMContextDispose(ctx);

  return 0;
}

File: tests/test_types.py
#!/usr/bin/env -S uv run
"""
Test: test_types
Tests LLVM Type creation and inspection via Python bindings.

This is the Python equivalent of tests/test_types.cpp.
Output should match the C++ golden master test.

LLVM APIs covered (via Python bindings):
- Integer types: ctx.types.i1, ctx.types.i8, ctx.types.i16, ctx.types.i32, ctx.types.i64, ctx.types.i128, ctx.types.int_n()
- Floating point: ctx.types.f16, ctx.types.bf16, ctx.types.f32, ctx.types.f64
- Other types: ctx.types.void, ctx.types.ptr, ctx.types.array(), ctx.types.vector(), ctx.types.function()
- Struct types: ctx.types.struct(), ctx.types.opaque_struct(), set_body()
- Type inspection: kind, int_width, is_sized, is_packed_struct, is_opaque_struct, struct_name
"""

import sys
import llvm


def type_kind_name(kind: llvm.TypeKind) -> str:
    """Convert TypeKind enum to the string name used in C++ test."""
    mapping = {
        llvm.TypeKind.Void: "void",
        llvm.TypeKind.Half: "half",
        llvm.TypeKind.Float: "float",
        llvm.TypeKind.Double: "double",
        llvm.TypeKind.FP128: "fp128",
        llvm.TypeKind.Label: "label",
        llvm.TypeKind.Integer: "integer",
        llvm.TypeKind.Function: "function",
        llvm.TypeKind.Struct: "struct",
        llvm.TypeKind.Array: "array",
        llvm.TypeKind.Pointer: "pointer",
        llvm.TypeKind.Vector: "vector",
        llvm.TypeKind.Metadata: "metadata",
        llvm.TypeKind.Token: "token",
        llvm.TypeKind.ScalableVector: "scalable_vector",
        llvm.TypeKind.BFloat: "bfloat",
    }
    return mapping.get(kind, "unknown")


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_types") as mod:
            # Integer types
            i1 = ctx.types.i1
            i8 = ctx.types.i8
            i16 = ctx.types.i16
            i32 = ctx.types.i32
            i64 = ctx.types.i64
            i128 = ctx.types.i128
            i256 = ctx.types.int_n(256)

            # Floating point types
            f16 = ctx.types.f16
            bf16 = ctx.types.bf16
            f32 = ctx.types.f32
            f64 = ctx.types.f64

            # Void type
            void_ty = ctx.types.void

            # Pointer type (opaque pointer)
            ptr = ctx.types.ptr

            # Array type
            arr_i32_10 = i32.array(10)

            # Vector type
            vec_i32_4 = i32.vector(4)

            # Function type: i32(i32, i32)
            func_ty = ctx.types.function(i32, [i32, i32], vararg=False)

            # Function type with varargs: i32(i32, ...)
            vararg_func_ty = ctx.types.function(i32, [i32], vararg=True)

            # Anonymous struct type: {i32, f64}
            anon_struct = ctx.types.struct([i32, f64], packed=False)

            # Packed anonymous struct type: <{i8, i32}>
            packed_struct = ctx.types.struct([i8, i32], packed=True)

            # Named struct type
            named_struct = ctx.types.opaque_struct("MyStruct")
            named_struct.set_body([i32, ptr, f64], packed=False)

            # Opaque struct (no body set)
            opaque_struct = ctx.types.opaque_struct("OpaqueStruct")

            # Add global variables to make types visible in output
            mod.add_global(i32, "global_i32")
            mod.add_global(arr_i32_10, "global_arr")
            mod.add_global(vec_i32_4, "global_vec")
            mod.add_global(anon_struct, "global_anon_struct")
            mod.add_global(packed_struct, "global_packed_struct")
            mod.add_global(named_struct, "global_named_struct")

            # Add function declarations to show function type
            mod.add_function("example_func", func_ty)
            mod.add_function("example_vararg_func", vararg_func_ty)

            # Verify module
            if not mod.verify():
                print(
                    f"; Verification failed: {mod.get_verification_error()}",
                    file=sys.stderr,
                )
                return 1

            # Print diagnostic comments
            print("; Test: test_types")
            print(";")
            print("; Integer types:")
            print(f";   i1 width: {i1.int_width}, kind: {type_kind_name(i1.kind)}")
            print(f";   i8 width: {i8.int_width}, kind: {type_kind_name(i8.kind)}")
            print(f";   i16 width: {i16.int_width}, kind: {type_kind_name(i16.kind)}")
            print(f";   i32 width: {i32.int_width}, kind: {type_kind_name(i32.kind)}")
            print(f";   i64 width: {i64.int_width}, kind: {type_kind_name(i64.kind)}")
            print(
                f";   i128 width: {i128.int_width}, kind: {type_kind_name(i128.kind)}"
            )
            print(
                f";   i256 width: {i256.int_width}, kind: {type_kind_name(i256.kind)}"
            )
            print(";")
            print("; Floating point types:")
            print(f";   half kind: {type_kind_name(f16.kind)}")
            print(f";   bfloat kind: {type_kind_name(bf16.kind)}")
            print(f";   float kind: {type_kind_name(f32.kind)}")
            print(f";   double kind: {type_kind_name(f64.kind)}")
            print(";")
            print("; Other types:")
            print(
                f";   void kind: {type_kind_name(void_ty.kind)}, sized: {'yes' if void_ty.is_sized else 'no'}"
            )
            print(
                f";   pointer kind: {type_kind_name(ptr.kind)}, sized: {'yes' if ptr.is_sized else 'no'}"
            )
            print(
                f";   array kind: {type_kind_name(arr_i32_10.kind)}, sized: {'yes' if arr_i32_10.is_sized else 'no'}"
            )
            print(
                f";   vector kind: {type_kind_name(vec_i32_4.kind)}, sized: {'yes' if vec_i32_4.is_sized else 'no'}"
            )
            print(
                f";   function kind: {type_kind_name(func_ty.kind)}, sized: {'yes' if func_ty.is_sized else 'no'}"
            )
            print(";")
            print("; Struct types:")
            print(
                f";   anon_struct kind: {type_kind_name(anon_struct.kind)}, packed: {'yes' if anon_struct.is_packed_struct else 'no'}"
            )
            print(
                f";   packed_struct kind: {type_kind_name(packed_struct.kind)}, packed: {'yes' if packed_struct.is_packed_struct else 'no'}"
            )

            named_struct_name = named_struct.struct_name
            print(
                f";   named_struct name: {named_struct_name}, opaque: {'yes' if named_struct.is_opaque_struct else 'no'}"
            )

            opaque_struct_name = opaque_struct.struct_name
            print(
                f";   opaque_struct name: {opaque_struct_name}, opaque: {'yes' if opaque_struct.is_opaque_struct else 'no'}"
            )

            # Print type strings
            print(";")
            print("; Type strings:")
            print(f";   i32: {i32}")
            print(f";   [10 x i32]: {arr_i32_10}")
            print(f";   func type: {func_ty}")

            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    sys.exit(main())
